design-proposals/configurable-features.md

```diff
@@ -0,0 +1,155 @@
+
+Development contributors.
+
+Cluster administrators.
+
+## User Stories
+
+As a developer, I want to make a given feature configurable.
+
+As a cluster administrator, I want to be able to change the cluster wide state of a feature by editing configurables.
+
+As VM owner, I want to use a given feature.
+
+## Repos
+
+Kubevirt/Kubevirt
+
+# Design
+
+If a developer wants to make a feature configurable, he needs to do so by adding new fields to the KubeVirt CR
+under `spec.configuration`.
+
+
+> **NOTE:** The inclusion of these new KubeVirt API fields should be carefully considered and justified. The feature
+> configurables should be avoided as much as possible.
+
+
+This is current list of GA'd features present in KubeVirt/KubeVirt which are still using feature gates and are shown as
+configurables in [HCO](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/controllers/operands/kubevirt.go#L166-L174):
+
+- DownwardMetrics
+- Root (not sure about this one)
+- DisableMDEVConfiguration
+- PersistentReservation
+- AutoResourceLimitsGate
+- AlignCPUs
+
+This is the current list of GA'd features present in KubeVirt/KubeVirt which are still using feature gates and are always
+enabled by [HCO](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/controllers/operands/kubevirt.go#L125-L142):
+
+- CPUManager
+- Snapshot
+- HotplugVolumes
+- GPU
+- HostDevices
+- NUMA
+- VMExport
+- DisableCustomSELinuxPolicy
+- KubevirtSeccompProfile
+- HotplugNICs
+- VMPersistentState
+- NetworkBindingPlugins
+- VMLiveUpdateFeatures
+
+Please note that only feature gates included in KubeVirt/KubeVirt are listed here.
+
+## API Examples
+The proposal configuration field, for a given feature in the KubeVirt CR, may look like:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+[...]
+spec:
+  certificateRotateStrategy: {}
+  configuration:
+    feature-A:  {}
+[...]
+```
+The VM object may or may not include a configuration field inside the relevant spec.
+
+## Interactions with the VMis requests
+
+In case that, the VM exposes a configuration field to request the feature as well as the KubeVirt CRD, the system may
+encounter some inconsistent states that should be handled in the following way:
+
+- If the feature is set to state A in the KubeVirt CR and the VMi is requesting the feature in state B, the VMis must
+  stay in Pending state. The VM status should be updated, showing a status message, highlighting the reason(s) for the
+  Pending state.
+- Feature status checks should only be performed during the scheduling process, not at runtime. Therefore, the feature
+  status changes in the KubeVirt CR should not affect running VMis. Moreover, the VMi should still be able to live
+  migrate, preserving its original feature state.
+- Optionally, It could enable the possibility to reject the KubeVirt CR change request if running VMis are using the
+  feature in a given state. However, by the default the request should be accepted.
```

Comment:
I suggested this use case and agree that as written it can easily race if a creation request from a user comes in while the admin is attempting to disable a given feature.

That said I think that at a high level the use case is still valid and that some cluster admins aren't going to want to disable a feature cluster wide that's being used by running VMs.

Reply:
> That said I think that at a high level the use case is still valid and that some cluster admins aren't going to want to disable a feature cluster wide that's being used by running VMs.

Right. I think that cluster admins would want to know if the functionality that they are disabling is currently being used. They may also want to know by whom (as not all VMs are created equal). But a mere VM owner should not be in the position to block the cluster admin from expressing an intent that a feature is to be reconfigured.

Reply:
I'm just a bit concerned about the time and resources needed to fetch all VMI using the feature.

Reply:
> > That said I think that at a high level the use case is still valid and that some cluster admins aren't going to want to disable a feature cluster wide that's being used by running VMs.
> 
> Right. I think that cluster admins would want to know if the functionality that they are disabling is currently being used. They may also want to know by whom (as not all VMs are created equal). But a mere VM owner should not be in the position to block the cluster admin from expressing an intent that a feature is to be reconfigured.

+1
I'm not sure we can make a simplistic rule here, it sounds possible that in some cases it's useful to change configuration for running VMs and in others it isn't. Perhaps it's useful to start off with more concrete examples and have a discussion regarding them.

Reply:
> I'm just a bit concerned about the time and resources needed to fetch all VMI using the feature.

Would you elaborate, possibly with a specific example? Would it be much harder than `kubectl get vm -o yaml|grep someRequestedFeature`?

+1 for considering several concrete examples.

Reply:
No, it won't be much harder. However, from my limited point of view, that would imply to fetch each VM in the system and check if the feature is present. That would result in an algorithmic complexity O(n). I'm not sure if such complexity is acceptable. Maybe there is a fancier way to do this that I'm not aware of.

Regarding of changing configurations in running VMs: I've proposed to allow changing configurations that do not require a VM restart to take effect, e.g., `EvictionStrategy`.

Reply:
This complexity is paid by the human who is responsible to make the cluster-wide changes. I don't think there's any way around it. The cluster-admin must be familiar with the workload and whether the feature he or she tweaks is needed now or in the near future.

>  I've proposed to allow changing configurations that do not require a VM restart to take effect, e.g., EvictionStrategy

Do you suggest to disallow modification of the kubevirt cr? Isn't it against https://github.com/kubevirt/community/pull/316/files#diff-48d9142aa505032f3643eef6ca3fb52e36bf8e84608a9649565619133f07490dR61 ?

Reply:
> This complexity is paid by the human who is responsible to make the cluster-wide changes. I don't think there's any way around it. The cluster-admin must be familiar with the workload and whether the feature he or she tweaks is needed now or in the near future.

Therefore, in your opinion, KubeVirt should warn the admin if there's affected VMs by the change, right? But in any means, we shouldn't block the update.

> Do you suggest to disallow modification of the kubevirt cr? Isn't it against https://github.com/kubevirt/community/pull/316/files#diff-48d9142aa505032f3643eef6ca3fb52e36bf8e84608a9649565619133f07490dR61 ?

No. At least by default, any modification should be allowed in the kubevirt CR.

Reply:
> Therefore, in your opinion, KubeVirt should warn the admin if there's affected VMs by the change, right? But in any means, we shouldn't block the update

Correct. And in "warning" I means "raise and alert or event or condition" on any affected VMs.

> No. At least by default, any modification should be allowed in the kubevirt CR.

Good. But we should not have a non-default attempt to block modification of the spec. It cannot be done safely (it is raceful), so we should not try.

Reply:
> Correct. And in "warning" I means "raise and alert or event or condition" on any affected VMs.

Ah I see, so should be the VMs itself who triggers the alert, event, warning, etc. Did I understand correctly?

> Good. But we should not have a non-default attempt to block modification of the spec. It cannot be done safely (it is raceful), so we should not try.

Works for me.

Reply:
> should be the VMs itself who triggers the alert, event, warning, etc.

As a VM owner, I see value in knowing that my VM is using a feature that is being removed from the cluster. So yes, a VM condition would be useful.

Reply:
All right. Would that also work for the cluster admin changing configurations? I.e., as a cluster admin, I'm changing configurables and want to know if there's any affected VM? or should be the cluster admin careful enough to run a command the one you have posted before `kubectl get vm -A -o yaml|grep someRequestedFeature` to know it in advance?

Reply:
> should be the cluster admin careful enough to run a ...  `kubectl get vm -A -o yaml|grep someRequestedFeature` to know it in advance?

I don't see any other way. Note that this is also raceful. Someone may create a VM with `someRequestedFeature` a microsecond later. That's the nature of our platform - the cluster admin cannot "know" at any point. They can only declare what is desired, and eventually the cluster would get there.

Reply:
All right. Thanks for the input. I will drop the sentence or change it by something like "the kubevirt CR can't be rejected, the cluster admin is responsible to know in advance if the feature configurable change will affect running VMIs".